spec/services/reports/monthly_bulk_report_service_spec.rb

```diff
@@ -5,14 +5,45 @@
 module Reports
   RSpec.describe MonthlyBulkReportService do
     describe ".run" do
-      it "generates a CSV file containing exemptions for the given month" do
-        monthly_bulk_serializer = double(:monthly_bulk_serializer)
+      let(:first_day_of_the_month) { Date.new(2019, 6, 1) }
+      context "when the request succeed" do
+        it "generates a CSV file containing exemptions for the given months and upload it to AWS" do
+          stub_successful_request
+          create_list(:registration_exemption, 2, :with_registration, registered_on: first_day_of_the_month)
 
-        expect(MonthlyBulkSerializer).to receive(:new).and_return(monthly_bulk_serializer)
-        expect(monthly_bulk_serializer).to receive(:to_csv)
+          expect_no_error
 
-        MonthlyBulkReportService.run(Date.today)
+          MonthlyBulkReportService.run(first_day_of_the_month)
+        end
       end
+
+      context "when the request succeed" do
+        it "fails gracefully and report the error" do
+          stub_failing_request
+          create_list(:registration_exemption, 2, :with_registration, registered_on: first_day_of_the_month)
+
+          expect(Airbrake).to receive(:notify).once
+
+          MonthlyBulkReportService.run(first_day_of_the_month)
+        end
+      end
+    end
+
+    def stub_successful_request
+      stub_request(:put, %r{https:\/\/.*\.s3\.eu-west-1\.amazonaws\.com\/20190601-20190630\.csv.+})
+    end
+
+    def stub_failing_request
+      stub_request(
+        :put,
+        %r{https:\/\/.*\.s3\.eu-west-1\.amazonaws\.com\/20190601-20190630\.csv.+}
+      ).to_return(
+        status: 403
+      )
+    end
+
+    def expect_no_error
```

Comment:
Are you expecting to need this again in subsequent tests? Just querying if it needs to be in a separate method just yet. There seems to be only one call to it, and when we need the negative version we just put the `expect` straight into the test (line 25) rather than also use a method.

Reply:
I went for a separate method to avoid adding a comment to the code. Is confusing to see that expectation in line with the test without knowing why we check it. I basically went for the method for documentation purposes.

Reply:
Will move it back in the test and add a comment if you prefer that 🙇‍♀

Reply:
Its more in the success test we call this method which is actually just the line

```ruby
expect(Airbrake).to_not receive(:notify)
```

However in the failing test rather than call a method there is just the line

```ruby
expect(Airbrake).to receive(:notify).once
```

So the documentation argument makes sense, but then I guess I would expect to see the failing test to be updated as

```ruby
it "fails gracefully and report the error" do
  ...
  expect_an_error
  ...
end

def expect_no_error
  expect(Airbrake).to_not receive(:notify)
end

def expect_an_error
  expect(Airbrake).to receive(:notify).once
end
```

Does this make sense? 😕

Reply:
yep it was inconsistent :/ I went for comments, that's more in line with the rest we have and doesn't make a huge difference